include/circt/Dialect/FIRRTL/FIRRTLInterfaces.td

```diff
@@ -0,0 +1,70 @@
+//===- FIRRTLInterfaces.td - FIRRTL op interfaces ----------*- tablegen -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This describes the MLIR interfaces for FIRRTL operations.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef FIRRTL_INTERFACES_TD
+#define FIRRTL_INTERFACES_TD
+
+include "mlir/IR/OpBase.td"
+
+// Op result type is entirely defined by its arguments.
+def ComputedType : OpInterface<"ComputedType"> {
+  let description = [{
+    An interface to operations whose result type is entirely defined by their
+    arguments (operand types and attributes).
+  }];
+  let methods = [
+
+    InterfaceMethod<[{
+      This method computes the result type of the operation from its arguments.
+      Operations must implement a `getResultType` function which actually
+      determines the type:
+
+      ```c++
+      class ConcreteOp ... {
+      protected:
+        FIRRTLType getResultType() { ... }
+      };
+      ```
+    }], "FIRRTLType", "getResultType", (ins), [{
```

Comment:
I think maybe calling it `computeResultType()` or something would make sense.  It seems a little confusing to me to have `op.getResultType()` and `op.getType()` which mean totally different things.

Reply:
Agreed! I think there are some places where MLIR uses "derive" for this deduction of types from operands. Maybe renaming this to "derivedType", "applyDerivedType", "verifyDerivedType"? But I'm happy with "computeResultType" as well. Anything that increases the contrast to "getType" 😉

Reply:
+1 for `computeResultType`